### PR TITLE
Add memory info to futhark bench json output

### PR DESCRIPTION
**Pull request by Munksgaard**
This commit adds support for benchmarking memory usage. It does so by changing `futhark bench` to always run with `-P`, and parses the output, searching for memory information. If any is found, it is added to the json output.

I've also added support in tools/cmp-bench-json.py for reporting the memory usage improvements, if present and significant.

screenshot

### CODE REVIEW

**Comment by athas** [comment]
I worry that using `-P` will reduce performance.  Is that the case?  We can always add another option that makes it print the memory statistics (which are unconditionally collected in all cases) but not enable all the other profiling instrumentation.  Maybe `-L` already does that.

**Comment by Munksgaard** [comment]
> I worry that using `-P` will reduce performance. Is that the case?

A quick test on OptionPricing.fut indicates that it does look to be the case for the CUDA backend, but not for the OpenCL backend.

I see you argument about adding another option instead. I think I'll do that instead.

**Comment by Munksgaard** [comment]
Is it okay if the log output comes after the program output? As in:

```
$ cat futhark-benchmarks/finpar/OptionPricing-data/small.in | futhark-benchmarks/finpar/OptionPricing -L
Using platform: NVIDIA CUDA
Using device: GeForce RTX 2080 Ti
Lockstep width: 32
Default group size: 256
Default number of groups: 272
Compared main.suff_outer_par_1 <= 1: false.
[167.055695f32]
Peak memory usage for space 'device': 3621040 bytes.
```

**Comment by athas** [comment]
As long as the memory logging is on stderr, that's fine.

And I'm actually going to replace this entire mechanism in #1179.

File: src/Futhark/CLI/Bench.hs [hunk 1]
```diff
@@ -297,7 +298,22 @@ runBenchmarkCase opts futhark program entry pad_to tr@(TestRun _ input_spec (Suc
...
+          Nothing ->
+            Left "Could not parse memory usage from profiling output"
+        )
+        & DataResult dataset_desc
+        & Just
+        & return
+
+getMemoryUsage :: T.Text -> Maybe Int
+getMemoryUsage t =
+  let mem_regex = "Peak memory usage for space '[^']+': ([0-9]+) bytes." :: T.Text
```
Comment (athas) [inline src/Futhark/CLI/Bench.hs:313]: What happens if there are multiple memory spaces?

Comment (Munksgaard) [reply src/Futhark/CLI/Bench.hs:313]: I am not really sure what this means, but cf. https://github.com/diku-dk/futhark/pull/1215/files/b4cb1dc490436b353326ca8268d1cde447a7961c#diff-c8ca25d5bf1f5928bbd214a58075cdf189a2dbbdee9007c06fceb8225d4238b0R315-R316, we will return `Nothing`. What would you prefer to happen?

Comment (athas) [reply src/Futhark/CLI/Bench.hs:313]: This should return a mapping from memory space to the peak, and the JSON should contain a dict.

We will certainly use both CPU and GPU memory in the same program in the relatively near future, and we might do even more funky things in the distant future, who knows.

**Comment by Munksgaard** [comment]
I've changed the implementation to return a map and dict instead. This is what it looks like with more than one device:

screenshot

File: src/Futhark/Bench.hs [hunk 2]
```diff
@@ -24,6 +24,7 @@ import qualified Data.Aeson as JSON
 import qualified Data.ByteString.Char8 as SBS
 import qualified Data.ByteString.Lazy.Char8 as LBS
 import qualified Data.HashMap.Strict as HM
+import Data.Map (Map)
```
Comment (athas) [inline src/Futhark/Bench.hs:27]: We generally use the qualified type name for this (`M.Map`).

File: src/Futhark/Bench.hs [hunk 3]
```diff
@@ -40,10 +41,10 @@ import System.Timeout (timeout)
 newtype RunResult = RunResult {runMicroseconds :: Int}
   deriving (Eq, Show)
 
--- | The results for a single named dataset is either an error
--- message, or runtime measurements along the stderr that was
+-- | The results for a single named dataset is either an error message, or
+-- runtime measurements, the number of bytes used, and the stderr that was
 -- produced.
-data DataResult = DataResult String (Either T.Text ([RunResult], T.Text))
+data DataResult = DataResult String (Either T.Text ([RunResult], Map T.Text Int, T.Text))
```
Comment (athas) [inline src/Futhark/Bench.hs:47]: Maybe it is time to actually create a data type for this triple.

File: unittests/Futhark/BenchTests.hs [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 module Futhark.BenchTests (tests) where
 
+import qualified Data.Map as Map
```
Comment (athas) [inline unittests/Futhark/BenchTests.hs:5]: `as M`!

**Review by athas — CHANGES_REQUESTED** [review]

**Comment by flip111** [comment]
Really cool feature @Munksgaard

File: src/Futhark/Bench.hs (hunk 3 above)
Comment (Munksgaard) [reply src/Futhark/Bench.hs]: Done. Although, I couldn't really think of a good name... Any suggestions?

**Review by athas — APPROVED** [review]

**Pull request merged**